ResearchKit/ActiveTasks/ORKHeartRateCameraRecorder.m

```diff
@@ -0,0 +1,609 @@
+    if (r == max) {
+        hue = (g - b) / delta;
+    } else if (g == max) {
+        hue = 2 + (b - r) / delta;
+    } else {
+        hue = 4 + (r - g) / delta;
+    }
+    hue *= 60;
+    if (hue < 0) {
+        hue += 360;
+    }
+    
+    *v = max;
+    *s = delta / max;
+    *h = hue;
+}
+
+- (NSArray *)butterworthBandpassFilter:(NSArray *)inputData {
+    const int NZEROS = 8;
+    const int NPOLES = 8;
+    static float xv[NZEROS+1], yv[NPOLES+1];
+    
+    // http://www-users.cs.york.ac.uk/~fisher/cgi-bin/mkfscript
+    // Butterworth Bandpass filter
+    // 4th order
+    // sample rate - varies between possible camera frequencies. Either 30, 60, 120, or 240 FPS
+    // corner1 freq. = 0.667 Hz (assuming a minimum heart rate of 40 bpm, 40 beats/60 seconds = 0.667 Hz)
+    // corner2 freq. = 4.167 Hz (assuming a maximum heart rate of 250 bpm, 250 beats/60 secods = 4.167 Hz)
+    // Bandpass filter was chosen because it removes frequency noise outside of our target range (both higher and lower)
+    double dGain = 1.232232910e+02;
+    
+    NSMutableArray *outputData = [[NSMutableArray alloc] init];
+    for (NSNumber *number in inputData) {
+        double input = number.doubleValue;
+        
+        xv[0] = xv[1]; xv[1] = xv[2]; xv[2] = xv[3]; xv[3] = xv[4]; xv[4] = xv[5]; xv[5] = xv[6]; xv[6] = xv[7]; xv[7] = xv[8];
+        xv[8] = input / dGain;
+        yv[0] = yv[1]; yv[1] = yv[2]; yv[2] = yv[3]; yv[3] = yv[4]; yv[4] = yv[5]; yv[5] = yv[6]; yv[6] = yv[7]; yv[7] = yv[8];
+        yv[8] = (xv[0] + xv[8]) - 4 * (xv[2] + xv[6]) + 6 * xv[4]
+        + ( -0.1397436053 * yv[0]) + (  1.2948188815 * yv[1])
+        + ( -5.4070037946 * yv[2]) + ( 13.2683981280 * yv[3])
+        + (-20.9442560520 * yv[4]) + ( 21.7932169160 * yv[5])
+        + (-14.5817197500 * yv[6]) + (  5.7161939252 * yv[7]);
+        
+        [outputData addObject:@(yv[8])];
+    }
+    
+    return outputData;
+}
+
+// Smoothed data helps remove outliers that may be caused by interference, finger movement or pressure changes.
+// This will only help with small interference changes.
+// This also helps keep the data more consistent.
+- (NSArray *)medianSmoothing:(NSArray *)inputData {
+    NSMutableArray *newData = [[NSMutableArray alloc] init];
+    
+    for (int i = 0; i < inputData.count; i++) {
+        if (i == 0 ||
+            i == 1 ||
+            i == 2 ||
+            i == inputData.count - 1 ||
+            i == inputData.count - 2 ||
+            i == inputData.count - 3)        {
+            [newData addObject:inputData[i]];
+        } else {
+            NSArray *items = [@[
+                                inputData[i-2],
+                                inputData[i-1],
+                                inputData[i],
+                                inputData[i+1],
+                                inputData[i+2],
+                                ] sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"self" ascending:YES]]];
```

Comment:
yes, how well it works is up for some debate. That's why there are two different methods included.